File: content/01-arrays-and-hashing/move_negatives.js
/**
 * Problem: Move Negative Numbers to Beginning
 * GeeksforGeeks: https://www.geeksforgeeks.org/move-negative-numbers-beginning-positive-end-constant-extra-space/
 */

function moveNegatives(arr) {
    /**
     * Moves all negative elements to the beginning of the array in-place.
     * Time Complexity: O(N)
     * Space Complexity: O(1)
     */
    let left = 0;
    for (let right = 0; right < arr.length; right++) {
        if (arr[right] < 0) {
            const temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
        }
    }
    return arr;
}

if (require.main === module) {
    const assert = require('assert');
    const arr1 = [-12, 11, -13, -5, 6, -7, 5, -3, -6];
    moveNegatives(arr1);
    
    // Validate first 6 elements are negative and the rest positive
    for (let i = 0; i < 6; i++) assert.ok(arr1[i] < 0);
    for (let i = 6; i < arr1.length; i++) assert.ok(arr1[i] >= 0);
    console.log("JS: moveNegatives passed tests!");
}

File: content/01-arrays-and-hashing/move_negatives.py
"""
Problem: Move Negative Numbers to Beginning
GeeksforGeeks: https://www.geeksforgeeks.org/move-negative-numbers-beginning-positive-end-constant-extra-space/
"""

def move_negatives(arr: list[int]) -> list[int]:
    """
    Moves all negative elements to the beginning of the array in-place.
    Time Complexity: O(N)
    Space Complexity: O(1)
    """
    left = 0
    for right in range(len(arr)):
        if arr[right] < 0:
            arr[left], arr[right] = arr[right], arr[left]
            left += 1
    return arr

if __name__ == "__main__":
    arr1 = [-12, 11, -13, -5, 6, -7, 5, -3, -6]
    move_negatives(arr1)
    # Check that all elements up to index 5 are negative
    assert all(x < 0 for x in arr1[:6])
    assert all(x >= 0 for x in arr1[6:])
    print("Python: move_negatives passed tests!")
